rfc/reproducible-builds.md

```diff
@@ -0,0 +1,66 @@
+# Reproducible Builds
+To ensure the security of constellation's supply chain, we need to make our software builds reproducible.
+This is the only way to ensure a verifiable path from source code to binary.
+Every step of the build process must be deterministic.
+
+## Definition of Goals
+1. All our OCIs executed in the cluster must be reproducible bit by bit .
+1. For that, every compiled executable has to be deterministically compiled.
+1. If we change parts of our codebase (i.e. add several other programming languages) we should not have to make major changes to the image build system.
+1. Moving from `Docker` to a reproducible build system should have a minimum overhead. Ideally, we can reuse our existing `Dockerfile`s.
+1. The tool that builds the OCIs should be battle proven and reliable.
+1. The image should be lightweight.
+
+## Steps to Achieve Goals
+
+### Executables
+* Remove metadata from binaries, namely:
+  * Timestamps
+  * Build IDs
+  * Symbols
+  * Version numbers
+  * Paths
+  * Anything that might change for another host OS / time / ...
+* Eliminate dependencies on libraries (make executable static)
+
+Striping metadata from the binary can be done in the building process.
+This can be achieved by setting the appropriate compiler flags.
+
+```bash
+$ CGO_ENABLED=0 go build -o <out_name> -buildvcs=false -trimpath -ldflags "-s -w -buildid=''"
+```
+
+### OCIs
+For the OCIs to be deterministic, each component of the image has to be deterministic as well.
+This includes:
+* The base image used to build the software must be the same for each build of a version. Pin the version with its `sha256` hash checksum.
+* The timestamps of the files in the image (creation, modification) must be identical for each build.
+* Every component that is shipped with the image has to be identical.
+
+To achieve this we will use [buildah](https://github.com/containers/buildah). It currently meets all the requirements mentioned above.
+
+To ensure that the final image is deterministic, a pattern such as the following should be followed:
+
+```Dockerfile
+FROM <image>@sha256:<hash> as build
+RUN  <install_deps>
```

Comment:
How do we ensure that we get the same version from the remote repository? We can pin versions, but does Fedora actually keep older package versions on their repositories? Ubuntu, for example, does not.

Reply:
For any binary that does not use CGO, we should not have any dependency other than Go (and the tools required to install Go).
Go keeps older version available, the version of tools like wget does not matter for the final build, as they are not used by our binary.

Container images that need CGO will require us pinning the versions of the C libraries we link against. These are typically not updated frequently and I think Fedora keeps them around for longer. However, we obviously need to check this first.

Reply:
The CGo case is where I could see issues. From googling, it seems that Koji has longer term storage of packages, but older packages can indeed disappear on the mirrors.

But I guess we would need to test that ourselves to have a better answer than educated guessing.

Reply:
Maybe we can have images with explicitly two layers for CGO:

- All dependencies (shared libraries) on one layer
- Our go binary on the upper layer

This allows us to reuse the dependency layer from our own container registry in catastrophic events.
We could even split this into explicit Dockerfiles for external deps and our own code to make the layers addressable.

Reply:
> This allows us to reuse the dependency layer from our own container registry in catastrophic events. We could even split this into explicit Dockerfiles for external deps and our own code to make the layers addressable.

We chose this approach for EGo and EdgelessDB. (Just for reproducible build of the binaries. The container images are not reproducible.)